internal/orchestration/components.go

```diff
@@ -0,0 +1,468 @@
+	Config *netconfig.Config
+	Sender *sender.Sender
+}
+
+func NewNetworkComponents(
+	coreComps *CoreComponents,
+	creds credentials.TransportCredentials,
+	mgrOpts ...gorums.ManagerOption,
+) *NetworkComponents {
+	cfg := netconfig.NewConfig()
+	send := sender.New(
+		cfg,
+		coreComps.EventLoop,
+		coreComps.Logger,
+		coreComps.Options,
+		creds,
+		mgrOpts...,
+	)
+
+	return &NetworkComponents{
+		Config: cfg,
+		Sender: send,
+	}
+}
+
+type SecurityComponents struct {
+	BlockChain *blockchain.BlockChain
+	CryptoImpl modules.CryptoBase
+	CertAuth   *certauth.CertAuthority
+}
+
+func NewSecurityComponents(
+	coreComps *CoreComponents,
+	netComps *NetworkComponents,
+	cryptoName string,
+	cacheSize int,
+) (*SecurityComponents, error) {
+	blockChain := blockchain.New(
+		netComps.Sender,
+		coreComps.EventLoop,
+		coreComps.Logger,
+	)
+	cryptoImpl, err := getCrypto(cryptoName, netComps.Config, coreComps.Logger, coreComps.Options)
+	if err != nil {
+		return nil, err
+	}
+	var certAuthority *certauth.CertAuthority
+	if cacheSize > 0 {
+		certAuthority = certauth.NewCached(
+			cryptoImpl,
+			blockChain,
+			coreComps.Logger,
+			cacheSize,
+		)
+	} else {
+		certAuthority = certauth.New(
+			cryptoImpl,
+			blockChain,
+			coreComps.Logger,
+		)
+	}
+	return &SecurityComponents{
+		BlockChain: blockChain,
+		CryptoImpl: cryptoImpl,
+		CertAuth:   certAuthority,
+	}, nil
+}
+
+type ServiceComponents struct {
+	CmdCache  *clientsrv.CmdCache
+	ClientSrv *clientsrv.ClientServer
+	Committer *committer.Committer
+}
+
+func NewServiceComponents(
+	coreComps *CoreComponents,
+	secureComps *SecurityComponents,
+	batchSize int,
+	clientSrvOpts []gorums.ServerOption,
+) *ServiceComponents {
+	cmdCache := clientsrv.NewCmdCache(
+		coreComps.Logger,
+		batchSize,
+	)
+	clientSrv := clientsrv.NewClientServer(
+		coreComps.EventLoop,
+		coreComps.Logger,
+		cmdCache,
+		clientSrvOpts,
+	)
+	committer := committer.New(
+		secureComps.BlockChain,
+		clientSrv,
+		coreComps.Logger,
+	)
+	return &ServiceComponents{
+		CmdCache:  cmdCache,
+		ClientSrv: clientSrv,
+		Committer: committer,
+	}
+}
+
+type ProtocolModules struct {
+	ConsensusRules modules.ConsensusRules
+	LeaderRotation modules.LeaderRotation
+	ViewDuration   modules.ViewDuration
+}
+
+type ViewDurationOptions struct {
+	SampleSize   uint64
+	StartTimeout float64
+	MaxTimeout   float64
+	Multiplier   float64
+}
+
+func NewProtocolModules(
+	coreComps *CoreComponents,
+	netComps *NetworkComponents,
+	secureComps *SecurityComponents,
+	srvComps *ServiceComponents,
+
+	opts *orchestrationpb.ReplicaOpts, // TODO: avoid modifying this so it doesn't depend on orchestrationpb
+	consensusName, leaderRotationName, byzantineName string,
+	vdOpt ViewDurationOptions,
+) (*ProtocolModules, error) {
+	consensusRules, err := getConsensusRules(consensusName, secureComps.BlockChain, coreComps.Logger, coreComps.Options)
+	if err != nil {
+		return nil, fmt.Errorf("invalid consensus name: '%s'", consensusName)
```

Comment:
Should return `err` directly, since the same error message is returned from `getConsensusRules`.